update to latest release-2.17.1-grafana (#41)
grafana/agent: docs/maintaining.md

Update the release branch by fast-forwarding it to the appropriate SHA (matching
the latest tag) and pushing it back upstream.

> ## `grafana/prometheus` Maintainence

Reviewer: ```suggestion
## `grafana/prometheus` Maintenance
```

> Grafana Labs is using the Agent for their internal monitoring and want to take

Reviewer 2: This line makes it sound like we're not also deploying/monitoring via vanilla Prometheus, which we are :)

> Grafana Labs is using the Agent for their internal monitoring and want to take

Reviewer: ```suggestion
In addition to deploying Prometheus, Grafana Labs is using the Agent for their internal monitoring and want to take
```

advantage of the Agent to proof-of-concept additions to Prometheus before they
get moved upstream. A `grafana/prometheus` repository is maintained by Grafana
Labs where all non-trivial changes will go first. Doing this allows for getting

> Cortex-specific changes moving along faster and providing strong evidence

Reviewer 2: Are most changes we're testing in the agent first really Cortex specific? I'd say:
> Testing these more experimental changes in a separate repo. that we can vendor into the agent allows us to iterate on them faster and...

towards its usefulness and correctness when it becomes part of an upstream PR.

We are commiting ourselves to doing the following:

> 1. Always use a recent Prometheus release: the Agent will always vendor a

Reviewer 2: We should clarify that we're vendoring a branch that's based on the latest/a recent release.

   recent Prometheus release and not Prometheus master. We want the Agent's

>    Prometeheus roots to be stable.

Reviewer 2: mispelling here

>    Prometeheus roots to be stable.

Reviewer: ```suggestion
   Prometheus roots to be stable.
```

> 2. Keep changes mergeable upstream: we want to continue to be good OSS citizens,

Reviewer 2: put this first imo

   and we intend that all features we add to our Prometheus repository will
   become an upstream PR. We will maintain our repository in a way that supports
   doing this.
3. Reduce code drift: The code the Agent uses on top of Prometheus will be
   layered on top of a Prometheus release rather than sandwiched in between.

Maintainence of the `grafana/prometheus` repository revolves around feature
branches (named `feat-SOME-FEATURE`) and release branches (named
`release-vX.Y.Z-grafana`). The release branches will always use the same release
version as the `prometheus/prometheus` release it is based off of.

> By adding features to the `grafana/prometheus` repository first, we are

Reviewer 2: this should be a 4th item in the list maybe `we commit to keeping the # of features not yet upstreamed from the grafana/prometheus repo to a minimum`, we could talk about how many major features we consider to be reasonable at next weeks meeting

committing ourselves to extra maintenance of features that have not yet been
merged upstream. Feature authors will have to babysit their features to

> coordinate with the Prometheus release schedule to always be compatible. One the

Reviewer 3: ```suggestion
coordinate with the Prometheus release schedule to always be compatible. Once the
```

Reviewer 3: I'm not sure I follow that last sentence. Once the feature is merged upstream - the Prometheus team can sync changes? Who is the Prometheus team in this context and how would they sync changes?

Author: Right, I was trying to say that it's harder for us to sync with upstream breaking changes because it's out of sync with the breaking changes (i.e., it only breaks for us once we update to the latest release).

Once the change is merged upstream, it's less of a problem. A breaking change is applied to work across the whole repository at once.

I'll try to phrase it better here to make that more clear, thanks!

feature is merged upstream, this burden of maintaining is made easier as the
Prometheus team can more easily sync breaking changes across the whole
repository.

We are purposefully carrying this extra burden because we intend to ultimately
make Prometheus better and contribute all of our enhancements upstream. We want
to strive to benefit the Prometheus ecosystem at large.

> ### Creating a New Feature

Reviewer 2: I think we should mention that we only intend to create these feature branches after we've attempted to upstream the change and it's clear the feature is more experimental or unproven than upstream is comfortable with at the moment.

For `grafana/prometheus` maintainers to create a new feature, they will do the
following:

1. Create a feature branch in `grafana/prometheus` based on the latest release
   tag that `grafana/prometheus` currently has a release branch for. The feature
   branch should follow the naming convention `feat-<feature name>`.
2. Implement the feature and open a PR to merge the feature branch into the
   associated `grafana/prometheus` release branch.

> 3. Once the release branch is updated, open a PR to update `grafana/agent` to

Reviewer 3: How do you feel about listing the current version we're using somewhere? That way we can keep track of which branch should we currently using. The alternative is scanning through PRs to check which is latest as I _think_ go.mod won't list it in a human-readable format, right?

Reviewer 3: The main idea behind my comment is to eventually add: `to know which is the current branch and version we're tracking please refer to this link in the readme somewhere`

Author: This would be a good thing to include in the README along with our reasoning for not using `prometheus/prometheus` directly.

   use the latest release branch SHA.

### Updating an Existing Feature

If a feature branch that was already merged to a release branch needs to be
updated for any reason:

> 1. Push directly to the feature branch or open a PR to merge changes into that

Reviewer 3: I'm not sure about encouraging trunk-based development by pushing directly. Do you feel confident about this?

Author: I do, because review of changes that were pushed directly to the feature branch can still be made in the PR which merges it into the release branch.

Reviewer 3: Sounds good. My worry here was about the new changes you would push (if needed) - out of the two I would almost always prefer to open the PR unless it's an emergency and then I wouldn't even bother following _any_ protocol.

   feature branch.
2. Open a PR to merge the new changes from the feature branch into the
   associated release branch.

> 3. Once the release branch is updated, open a PR to update `grafana/agent` to
>    use the latest release branch SHA.

Reviewer 3: ```suggestion
3. Once the release branch is updated, open a PR to update `grafana/agent` by vendoring
   the changes using the latest release branch SHA.
```

### Handling New Upstream Release

When a new upstream `prometheus/prometheus` release is available, we must go
through the following process:

1. Create a new `grafana/prometheus` release branch named
   `release-X.Y.Z-grafana`.
2. For all feature branches still not merged upstream, rebase them on top of the

>    latest release. Force push them to update the `grafana/prometheus` release

Reviewer 3: ```suggestion
   newly created release. Force push them to update the `grafana/prometheus` release
```

Reviewer 3: Do we want to say release branch? I thought what we need is to update the feature branch and then merge those changes into the newly created release branch.

   branch.
3. Create one or more PRs to introduce the features into the newly created
   release branch.

> Once this process is completed, the previous release branch in

Reviewer 3: then.. X place which holds the current version should be updated.

`grafana/prometheus` is considered stale and will no longer be updated.

### Updating the Agent's vendor

The easiest way to do this is the following:

1. Edit `go.mod` and change the replace directive to the release branch name.

> 2. Run `go mod vendor && go mod tidy`.

Reviewer 3: Aren't we supposed to run `go mod tidy` before `go mod vendor`, I think we want to remove the unused dependencies before vendoring.

Author: Yep, nice catch!

3. Commit and open a PR.

> ### Gotchas

Reviewer 3: I know this section is all about options, but if possible (and it might not be) we should settle on one. Having multiple processes for conflicting branches can ruin the flow and make it harder on ourselves.

Ideally, I'd love to have people wait until a new release but that might not be feasible.

Author: Yeah, the second option is going to be hard to coordinate. Let me think about this a little more. I agree with settling on one preferred option, but I'd still like to outline the "in case of emergency, break glass" alternatives.

If the `grafana/prometheus` feature is incompatible with the upstream master

> branch, an upstream PR cannot be merged due to the merge conflicts. There are a

Reviewer 3: I'm not following here, we're using the word upstream twice in a different context - am I missing something?

few ways this can be handled at the feature author's discrection:

1. Rebase the feature branch to `prometheus/prometheus` master so it can be

>    merged upstream. Doing this means that we cannot have the feature in th agent

Reviewer 3: ```suggestion
   merged upstream. Doing this means that we cannot have the feature in the agent
```

   until a new upstream release is available containing the feature.
2. Wait until a new `prometheus/prometheus` release is available and rebase the
   feature branch on top. The upstream PR will now be compatible with master,
   but this window is small and may change at any time. This makes it slower to
   get the feature merged upstream.
3. Create a new feature branch based off of master and open a PR for that
   feature branch. This adds extra maintenance burden on the feature author as
   they now have to mirror changes across two feature branches.

### Open Questions

> If two feature branches depend on one another, there are two suggested solutions

Reviewer 3: Same as above, let's settle on one. I'm happy with the first if it feels simpler.
